mesonbuild/backend/ninjabackend.py

```diff
@@ -159,33 +159,41 @@ class Quoting(Enum):
     none = 3
 
 class NinjaCommandArg:
-    def __init__(self, s, quoting = Quoting.both):
+    def __init__(self, s: str, quoting: Quoting = Quoting.both) -> None:
         self.s = s
         self.quoting = quoting
 
-    def __str__(self):
+    def __str__(self) -> str:
         return self.s
 
     @staticmethod
-    def list(l, q):
+    def list(l: str, q: Quoting) -> T.List[NinjaCommandArg]:
         return [NinjaCommandArg(i, q) for i in l]
 
+@dataclass
 class NinjaComment:
-    def __init__(self, comment):
-        self.comment = comment
+    comment: str
 
-    def write(self, outfile):
+    def write(self, outfile: T.TextIO) -> None:
         for l in self.comment.split('\n'):
             outfile.write('# ')
             outfile.write(l)
             outfile.write('\n')
         outfile.write('\n')
 
+@dataclass
 class NinjaRule:
-    def __init__(self, rule, command, args, description,
-                 rspable = False, deps = None, depfile = None, extra = None,
-                 rspfile_quote_style: RSPFileSyntax = RSPFileSyntax.GCC):
-
+    name: str
+    command: T.List[T.Union[NinjaCommandArg, str]]
+    args: T.List[T.Union[NinjaCommandArg, str]]
```

Comment:
I don't believe this is actually allowed -- maybe with python 3.13 dataclasses.field(converter=) you can pass values to init that are wider than the class-wide annotation.

Without the ability to type-narrow I figured the complexity of InitVar wasn't worth it.

We could just forego dataclasses and keep a hand-written init method but add some missing typing. I'd be happy to do that too.

Reply:
I found some... interesting mypy behavior (which pyright also does), so it might not be a bug. The following works:

```python
@dataclass
class NinjaRule:

    command_: InitVar[T.Sequence[NinjaCommandArg | str]]
    args_: InitVar[T.Sequence[NinjaCommandArg | str]]

    def __post_init__(self, cmd: T.Sequence[NinjaCommandArg | str], args: T.Sequence[NinjaCommandArg | str]) -> None:
        def converter(thing: T.Sequence[NinjaCommandArg | str]) -> T.List[NinjaCommandArg]: ...
        if isinstance(cmd, str):
            cmd = [cmd]  # or assert, whatever
        if isinstance(args, str):
            args = [args]
        self.command = convertor(cmd)
        self.args = convertor(args)
```

The "interesting" behavior is that if the argument is called `command` not `command_` then mypy gets grumpy about you assigning things you shouldn't.

I also think it's worth handling Sequence here and the potential to get a `str`, because otherwise we'll end up having to cast every time we want to pass a `list[NinjaCommandArg]`, and memory tells me that there are cases where mypy is going to figure out that what we're passing is not a union.

Reply:
Has mypy.or the typing SIG solved the "sequence of str is equal to one str" issue yet?

Still waiting for typing.CovariantList[str] that accepts lists of str but not str and not tuples of str.

Also still waiting for typing.CovariantContainerizedHolder[str] that accepts two out of three for people that aren't picky about tuples of str.

Reply:
My solution here is to defer using dataclasses, and expect people to properly type their lists of failed covariants in order to properly typecheck and error out if someone tries to do:
```python
NinjaRule('FOO_NAME', 'mycmd', ...)
```

Reply:
sounds good to me